File: app/transform/frontend/src/services/cardService.js
import api from './api'

/**
 * Card Service
 * Handles all card-related API calls
 */

/**
 * Get card details by card number
 * @param {string} cardNumber - 16-digit card number
 * @returns {Promise<Object>} Card details
 */
export async function getCard(cardNumber) {
  try {
    const response = await api.get(`/cards/${cardNumber}`)
    return response.data
  } catch (error) {
    throw error
  }
}

/**
 * Get all cards for an account
 * @param {string} accountId - 11-digit account ID
 * @returns {Promise<Array>} List of cards
 */
export async function getCardsByAccount(accountId) {
  try {
    const response = await api.get(`/cards/account/${accountId}`)
    return response.data
  } catch (error) {
    throw error
  }
}

/**
 * Create a new card
 * @param {Object} cardData - Card creation data
 * @returns {Promise<Object>} Created card details
 */
export async function createCard(cardData) {
  try {
    const response = await api.post('/cards', cardData)
    return response.data
  } catch (error) {
    throw error
  }
}

/**
 * Update card information
 * @param {string} cardNumber - 16-digit card number
 * @param {Object} updates - Card update data
 * @returns {Promise<Object>} Updated card details
 */
export async function updateCard(cardNumber, updates) {
  try {
    const response = await api.put(`/cards/${cardNumber}`, updates)
    return response.data
  } catch (error) {
    throw error
  }
}

/**
 * Deactivate a card
 * @param {string} cardNumber - 16-digit card number
 * @returns {Promise<Object>} Deactivation confirmation
 */
export async function deactivateCard(cardNumber) {
  try {
    const response = await api.delete(`/cards/${cardNumber}`)
    return response.data
  } catch (error) {
    throw error
  }
}

export default {
  getCard,
  getCardsByAccount,
  createCard,
  updateCard,
  deactivateCard
}

File: app/transform/backend/src/main/java/com/carddemo/service/CardService.java
package com.carddemo.service;

import com.carddemo.dto.CardCreateDto;
import com.carddemo.dto.CardDto;
import com.carddemo.dto.CardUpdateDto;

import java.util.List;

/**
 * Service interface for Card management operations
 * Handles business logic for card viewing, creation, and updates
 */
public interface CardService {
    
    /**
     * Get card by card number
     * @param cardNumber the card number to retrieve
     * @return CardDto containing card details
     * @throws com.carddemo.exception.ResourceNotFoundException if card not found
     */
    CardDto getCard(Long cardNumber);
    
    /**
     * Get all cards for a specific account
     * @param accountId the account ID to retrieve cards for
     * @return List of CardDto objects
     */
    List<CardDto> getCardsByAccount(Long accountId);
    
    /**
     * Create a new card
     * @param cardData the card creation data
     * @return CardDto containing the created card details
     * @throws com.carddemo.exception.ValidationException if validation fails
     * @throws com.carddemo.exception.ResourceNotFoundException if account or customer not found
     */
    CardDto createCard(CardCreateDto cardData);
    
    /**
     * Update an existing card
     * @param cardNumber the card number to update
     * @param updates the card update data
     * @return CardDto containing the updated card details
     * @throws com.carddemo.exception.ResourceNotFoundException if card not found
     * @throws com.carddemo.exception.ValidationException if validation fails
     */
    CardDto updateCard(Long cardNumber, CardUpdateDto updates);
    
    /**
     * Deactivate a card (soft delete)
     * @param cardNumber the card number to deactivate
     * @throws com.carddemo.exception.ResourceNotFoundException if card not found
     */
    void deactivateCard(Long cardNumber);
}
